pi.py

```diff
@@ -1 +1,578 @@
+    RD=287.04        # [J/kg.K] gas constant of dry air
+    EPS=RD/RV        # [unitless] epsilon, the ratio of gas constants
+    ALV0=2.501e6     # [J/kg] latent heat of vaporization at 0 degrees C
+    
+    #
+    #  ***  Define various parcel quantities, including reversible   ***
+    #  ***                       entropy, S                          ***
+    #                         
+    TPC=TP-273.15                           # Parcel temperature in Celsius
+    ESP=6.112*np.exp(17.67*TPC/(243.5+TPC)) # Parcel's saturated vapor pressure
+    EVP=RP*PP/(EPS+RP)                      # Parcel's partial vapor pressure
+    RH=EVP/ESP                              # Parcel's relative humidity
+    RH=min([RH,1.0])                     # ensure that the relatively humidity does not exceed 1.0
+    ALV=ALV0+CPVMCL*TPC                     # calculate the latent heat of vaporization, dependant on temperature
+    # calculate reversible total specific entropy per unit mass of dry air (E94, EQN. 4.5.9)
+    S=(CPD+RP*CL)*np.log(TP)-RD*np.log(PP-EVP)+ALV*RP/TP-RP*RV*np.log(RH)
+    
+    #
+    #   ***  Estimate lifted condensation level pressure, PLCL   ***
+    #     Based on E94 "calcsound.f" code at http://texmex.mit.edu/pub/emanuel/BOOK/
+    #     see also https://psl.noaa.gov/data/composites/day/calculation.html
+    #
+    #   NOTE: Modern PLCL calculations are made following the exact expressions of Romps (2017),
+    #   see https://journals.ametsoc.org/doi/pdf/10.1175/JAS-D-17-0102.1
+    #   and Python PLCL code at http://romps.berkeley.edu/papers/pubdata/2016/lcl/lcl.py
+    #
+    CHI=TP/(1669.0-122.0*RH-TP)
+    PLCL=PP*(RH**CHI)
+    
+    # Initial default values before loop
+    CAPED=0
+    TOB=T[0]
+    IFLAG=1
+    # Values to help loop
+    NCMAX=0
+    jmin=int(1e6)
+    
+    #
+    #   ***  Begin updraft loop   ***
+    #
+
+    # loop over each level in the profile
+    for j in range(nlvl):
+        
+        # jmin is the index of the lowest pressure level evaluated in the loop
+        jmin=int(min([jmin,j]))
+    
+        #
+        #   *** Calculate Parcel quantities BELOW lifted condensation level   ***
+        #
+        if (P[j] >= PLCL):
+            # Parcel temperature at this pressure
+            TG=TP*(P[j]/PP)**(RD/CPD)
+            # Parcel Mixing ratio
+            RG=RP
+            # Parcel and Environmental Density Temperatures at this pressure (E94, EQN. 4.3.1 and 6.3.7)
+            TLVR=TG*(1.+RG/EPS)/(1.+RG)
+            TVENV=T[j]*(1.+R[j]/EPS)/(1+R[j])
+            # Bouyancy of the parcel in the environment (Proxy of E94, EQN. 6.1.5)
+            TVRDIF[j,]=TLVR-TVENV
+            
+        #
+        #   *** Calculate Parcel quantities ABOVE lifted condensation level   ***
+        # 
+        else:
+            
+            # Initial default values before loop
+            TGNEW=T[j]
+            TJC=T[j]-273.15
+            ES=6.112*np.exp(17.67*TJC/(243.5+TJC))
```

Comment:
I see you've calculated the saturation vapor pressure again. See the above comment on refactoring (you can easily reduce bugs this way!)

Reply:
Agreed.